internal/keepers/ocr.go

```diff
@@ -53,50 +55,73 @@ func (k *keepers) Query(_ context.Context, _ types.ReportTimestamp) (types.Query
 // Observation implements the types.ReportingPlugin interface in OCR2. This method samples a set
 // of upkeeps available in and UpkeepService and produces an observation containing upkeeps that
 // need to be executed.
-func (k *keepers) Observation(ctx context.Context, rt types.ReportTimestamp, _ types.Query) (types.Observation, error) {
-	lCtx := newOcrLogContext(rt)
-	ctx = context.WithValue(ctx, ocrLogContextKey{}, lCtx)
+func (k *keepers) Observation(ctx context.Context, reportTimestamp types.ReportTimestamp, _ types.Query) (types.Observation, error) {
+	lCtx := newOcrLogContext(reportTimestamp)
 
-	blockKey, results, err := k.service.SampleUpkeeps(ctx, k.filter.Filter())
-	if err != nil {
-		return nil, fmt.Errorf("%w: failed to sample upkeeps for observation: %s", err, lCtx)
+	if len(k.observers) == 0 {
+		return nil, fmt.Errorf("empty observer list: %s", lCtx)
 	}
 
-	// keyList produces a sorted result so the following reduction of keys
-	// should be more uniform for all nodes
-	keys := keyList(filterUpkeeps(results, ktypes.Eligible))
+	allIDs := make([]ktypes.UpkeepIdentifier, 0)
+	var blocks []*big.Int
 
-	obs := &chain.UpkeepObservation{
-		BlockKey: chain.BlockKey(blockKey.String()),
-	}
+	for _, observer := range k.observers {
+		block, ids, err := observer.Observe()
+		if err != nil {
+			return nil, fmt.Errorf("%w: failed to sample upkeeps for observation: %s", err, lCtx)
+		}
+
+		if block == nil {
+			k.logger.Printf("observed %d upkeep IDs with nil block", len(allIDs))
+		} else {
+			k.logger.Printf("observed %d upkeep IDs at block %s", len(allIDs), block)
+
+			if bigInt, ok := block.BigInt(); ok {
+				blocks = append(blocks, bigInt)
+			} else {
+				return nil, fmt.Errorf("%w: failed to parse block key for observation: %s", err, lCtx)
+			}
+		}
+
+		allIDs = append(allIDs, ids...)
 
-	identifiers := make([]ktypes.UpkeepIdentifier, 0)
-	for _, upkeepKey := range keys {
-		_, upkeepID, _ := upkeepKey.BlockKeyAndUpkeepID()
-		identifiers = append(identifiers, upkeepID)
 	}
 
-	// Shuffle the observations before we limit it to observationUpkeepsLimit
-	keyRandSource := getRandomKeySource(rt)
-	identifiers = shuffleObservations(identifiers, keyRandSource)
+	k.logger.Printf("observed a total of %d upkeep IDs", len(allIDs))
+
+	keyRandSource := getRandomKeySource(reportTimestamp)
+	allIDs = shuffleObservations(allIDs, keyRandSource)
+
 	// Check limit
-	if len(identifiers) > observationUpkeepsLimit {
-		identifiers = identifiers[:observationUpkeepsLimit]
+	if len(allIDs) > observationUpkeepsLimit {
+		allIDs = allIDs[:observationUpkeepsLimit]
+	}
+
+	medianBlock := calculateMedianBlock(blocks, k.reportBlockLag)
```

Comment:
not sure why we need this logic here? Is this the local aggregation of different observations here? max might make more sense then?

Reply:
specifically it does not make sense to subtract reportBlockLag here and changes the intended behaviour of this config

Reply:
We don't _really_ need a median block here, it's just that each observer returns a block, and so we have to go from having multiple blocks to having a single block, I can change this to use the zeroth block that gets returned? And that way we eliminate the median calculation and the use of reportBlockLag